external-crates/move/crates/move-regex-borrow-graph/src/references.rs

```diff
@@ -0,0 +1,439 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    Result, bail,
+    collections::{Path, Paths},
+    error,
+    regex::Regex,
+};
+use itertools::Either;
+use std::{
+    cmp,
+    collections::{BTreeMap, BTreeSet},
+    fmt::{self, Debug},
+};
+
+//**************************************************************************************************
+// Definitions
+//**************************************************************************************************
+
+/// A new type wrapper around Ref_ to not expose the internal variants.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct Ref(Ref_);
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd)]
+enum Ref_ {
+    /// A canonicalized reference--this lets join operate over the same domain
+    Canonical(usize),
+    /// A reference specific to this block
+    Fresh(usize),
+}
+
+#[derive(Clone)]
+struct LocRegex<Loc, Lbl> {
+    loc: Loc,
+    regex: Regex<Lbl>,
+}
+
+#[derive(Clone, Debug)]
+struct Edge<Loc, Lbl: Ord> {
+    abstract_size: usize,
+    regexes: BTreeSet<LocRegex<Loc, Lbl>>,
```

Comment:
Is there a reason this is a set instead of a location-keyed map? (I assume the answer is "yes" -- just trying to understand).

Reply:
Sigh cuz I'm dumb. 

In early versions I needed this type wrapper and never really rethought it. Thanks for the call out, I'll switch to map!